### PR TITLE
Run tests in different versions of PHP

### LINKED ISSUE #399: Run PHP unit tests in different versions of PHP
## Feature Description

Run PHP unit tests in different versions of PHP in CI.

### PR DESCRIPTION
**Pull request by mukeshpanchal27**
## Summary

Fixes #399

## Checklist

- [x] PR has either `[Focus]` or `Infrastructure` label.
- [x] PR has a `[Type]` label.
- [x] PR has a milestone or the `no milestone` label.

### CODE REVIEW

File: .github/workflows/php-test.yml [hunk 1]
```diff
@@ -39,9 +39,15 @@ on:
...
 jobs:
   php-test:
-    name: PHP
+    name: "PHP ${{ matrix.php }}"
     runs-on: ubuntu-latest
     timeout-minutes: 20
+    strategy:
+      fail-fast: false
+      matrix:
+        php: ['7.4','8.0','8.1']
```
Comment (swissspidy) [inline .github/workflows/php-test.yml]: Ideally we also test PHP 7.0, 8.2, and 8.3 (as experimental); not sure if wp-env supports all of these

**Comment by swissspidy** [comment]
I thought I'd pick this one up, but a couple of SVG upload tests (`test_get_dominant_color_none_images`) are failing on various PHP versions 🤷

**Comment by mukeshpanchal27** [comment]
Thanks @swissspidy for picking. I also don't understood why it going fails 🤔

**Comment by swissspidy** [comment]
So the failures makes sense, because WP doesn't handle/support SVG uploads, so `wp_check_filetype` fails. And unfiltered uploads are not enabled either. The weird part is that the test is passing on PHP 7.4...

We don't really need to test SVG upload here anyway, so we can just remove it.

**Comment by swissspidy** [comment]
Tests on PHP < 7.3 are failing because the PHPUnit version (v9.6) is too new for those PHP versions, causing syntax errors.

We probably don't need something as complex as https://github.com/WordPress/plugin-check/pull/321, but definitely something in that direction.

**Comment by swissspidy** [comment]
Just updated wp-env and force a `composer update` before running tests

Good news: PHPUnit is now running on every PHP version \*
Bad news: tests are failing because webp doesn't seem to be supported on these versions? 🤔 

\* Edit: need to move PHPStan to `build-cs` here to fix PHP 7.0 & 7.1

**Comment by swissspidy** [comment]
Okay so PHPUnit install is fixed now.

The last remaining issue seems to be that `wp_image_editor_supports( array( 'mime_type' => 'image/webp' ) )` in `webp_uploads_generate_additional_image_source` now suddenly seems to return `false`  on wp-env 🤔  That causes WebP uploads to fail in tests with `image_mime_type_not_supported` errors.

On the actual wp-env website the WebP uploads work just fine though, so no idea what's going on.

Unfortunately I am unable to get my Docker setup working right now so can't reproduce locally.

**Comment by swissspidy** [comment]
I really have no idea what's going on with these Imagick and WebP tests 🤷 😩

**Comment by swissspidy** [comment]
@felixarntz @adamsilverstein @spacedmonkey Has anyone got an idea why these dominant color and webp tests are suddenly failing with this PR? 🤷

**Comment by felixarntz** [comment]
@swissspidy Not sure that's what's causing the failures, but there's that hard-coded PHP 7.4 declaration in https://github.com/WordPress/performance/blob/trunk/.wp-env.json#L9. I wonder whether because of that the test suite is somehow on a different PHP version than core and therefore these problems are caused?

Just a hunch worth checking. I don't even know why this hard-coded version is there, so maybe we should try removing it and see what happens? 🤔

**Comment by swissspidy** [comment]
Doesn't seem to make a difference.

At least locally, I get the `it_should_validate_source_attribute_update_when_webp_edited` test to fail as well, even without wp-env.

As for dominant color, it's only the Imagick tests that fail. Maybe that's not properly available or so? 🤔

**Comment by mukeshpanchal27** [comment]
In https://github.com/WordPress/performance/pull/544/commits/baf7ce31678f73115b6ea4d8f00c7b42f8340c7b remove filter that disturbed other tests. now test suite show two error in PHP > 7.4 but for lower version it show error for more tests 🤔

**Comment by mukeshpanchal27** [comment]
@swissspidy Unit tests are now pass ✅ 7.4 and above PHP version. For lower PHP version we have to check what was missing.

**Comment by swissspidy** [comment]
On PHP < 7.4 it seems that there is no built-in WebP support or so.

**Comment by mukeshpanchal27** [comment]
> On PHP < 7.4 it seems that there is no built-in WebP support or so.

Yes. In https://github.com/WordPress/performance/pull/544/commits/e2872b8779347e862b97bc551d070f7f3cb78bbc I fid Unit test issue for PHP 7.3

**Comment by mukeshpanchal27** [comment]
We also need to update the command here: https://github.com/WordPress/performance/blob/fix/399-unit-tests-in-php-versions/bin/plugin/commands/test-plugins.js#L323-L343 so that the `test-plugins` script works as expected.

**Comment by swissspidy** [comment]
> We also need to update the command here: [`fix`/399-unit-tests-in-php-versions/bin/plugin/commands/test-plugins.js#L323-L343](https://github.com/WordPress/performance/blob/fix/399-unit-tests-in-php-versions/bin/plugin/commands/test-plugins.js?rgh-link-date=2024-01-24T12%3A30%3A36Z#L323-L343) so that the `test-plugins` script works as expected.

First we need to update `.github/workflows/php-test-standalone-plugins.yml` to use a matrix.

I'll do that now

**Comment by swissspidy** [comment]
Unit tests pass now 🎉 

Not sure how the integration tests are setup 🤔  @mukeshpanchal27 perhaps you can take a look at fixing that setup?

**Comment by mukeshpanchal27** [comment]
All the unit test pass ✅ now. Looking into integration tests setup.

**Comment by swissspidy** [comment]
There is no error output for failing runs, see e.g. https://github.com/WordPress/performance/actions/runs/7650368243/job/20846313421?pr=544

Do we need to log `command.stderr`?

**Comment by mukeshpanchal27** [comment]
@swissspidy Added log in script. In local i got below error for PHP 7.2.

```
ℹ Starting 'vendor/bin/phpunit -c phpunit.xml --verbose --testdox' on the tests-cli container. 
This version of PHPUnit requires PHP >= 7.3.
You are using PHP 7.2.34 (/usr/local/bin/php).
```

**Comment by swissspidy** [comment]
> @swissspidy Added log in script. In local i got below error for PHP 7.2.
> 
> ```
> ℹ Starting 'vendor/bin/phpunit -c phpunit.xml --verbose --testdox' on the tests-cli container. 
> This version of PHPUnit requires PHP >= 7.3.
> You are using PHP 7.2.34 (/usr/local/bin/php).
> ```

Aah 💡 

Now it makes sense. We need to downgrade PHPUnit in this case using `composer update`, just like we do for unit tests.

Maybe 7d560aeb42b528eed8b1449174fc35d3e4cd795b works 🤞

**Comment by swissspidy** [comment]
Or maybe not 🤷

**Review by swissspidy — APPROVED** [review]
Yay 🤩🥳

File: build-cs/composer.json
Comment (westonruter) [inline build-cs/composer.json:1]: Perhaps a naïve question: what does the "CS" stand for in `build-cs`?

Comment (swissspidy) [reply build-cs/composer.json:1]: Coding standards

**Review by westonruter — APPROVED** [review]

File: composer.json [hunk 2]
```diff
@@ -12,35 +12,34 @@
...
+      "build-cs/vendor/bin/phpstan analyse --memory-limit=2048M -c phpstan.neon.dist"
+    ],
+    "format": [
+      "composer --working-dir=build-cs update --no-interaction",
+      "build-cs/vendor/bin/phpcbf --standard=phpcs.xml.dist --report-summary --report-source"
+    ],
+    "lint": [
+      "composer --working-dir=build-cs update --no-interaction",
+      "build-cs/vendor/bin/phpcs --standard=phpcs.xml.dist"
+    ],
```
Comment (westonruter) [inline composer.json:34]: I found that calling `composer update` here causes problems: https://github.com/WordPress/performance/issues/961

Proposed fix: https://github.com/WordPress/performance/pull/962

**Pull request merged**